src/main/java/com/atlassian/db/replica/api/AuroraLsnReplicaConsistency.java

```diff
@@ -0,0 +1,99 @@
+
+/**
+ * It remembers last write LSN (log sequence number) and compares it with LSN for subsequent reads to detect inconsistencies.
+ * If it cannot remember the LSN of last write, pessimistically assumes it's going to be inconsistent.
+ *
+ * @see <a href="https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/aurora-global-database-monitoring.html">Monitoring Aurora PostgreSQL-based Aurora global databases</a>
+ * <p>
+ * It's an PostgreSQL-based Aurora database engine specific implementation.
+ */
+public class AuroraLsnReplicaConsistency implements ReplicaConsistency {
+
+    private final Cache<Long> lastWrite;
+
+    public static class Builder {
+        private Cache<Long> lastWrite = new MonotonicMemoryCache<>();
+
+        /**
+         * @param lastWrite remembers last write
+         */
+        public Builder cacheLastWrite(Cache<Long> lastWrite) {
+            this.lastWrite = lastWrite;
+            return this;
+        }
+
+        /**
+         * @return consistency assuming that LSN (log sequence number) is greater or equal to LSN for last write (if known)
+         */
+        public AuroraLsnReplicaConsistency build() {
+            return new AuroraLsnReplicaConsistency(lastWrite);
+        }
+    }
+
+    private AuroraLsnReplicaConsistency(Cache<Long> lastWrite) {
+        this.lastWrite = lastWrite;
+    }
+
+    @Override
+    public void write(Connection main) {
+        try {
+            lastWrite.put(queryMainDbLsn(main));
+        } catch (Exception e) {
+            lastWrite.reset();
+        }
+    }
+
+    @Override
+    public boolean isConsistent(Supplier<Connection> replica) {
+        Optional<Long> maybeLastWrite = lastWrite.get();
+        if (!maybeLastWrite.isPresent()) {
+            return false;
+        }
+
+        long lastRefresh;
+        try {
+            lastRefresh = queryReplicaDbLsn(replica.get());
+        } catch (Exception e) {
+            return false;
+        }
+
+        return lastRefresh >= maybeLastWrite.get();
+    }
+
+    /**
+     * @return LSN (log sequence number) for the main database (the highest LSN)
+     */
+    private long queryMainDbLsn(Connection connection) throws Exception {
+        return queryLsn(connection, "SELECT MAX(durable_lsn) AS lsn FROM aurora_global_db_instance_status();");
+    }
+
+    /**
+     * @return LSN (log sequence number) for the most outdated replica database (the lowest LSN)
+     */
+    private long queryReplicaDbLsn(Connection connection) throws Exception {
+        return queryLsn(connection, "SELECT MIN(durable_lsn) AS lsn FROM aurora_global_db_instance_status();");
+    }
+
+    private long queryLsn(Connection connection, String rawSqlQuery) throws SQLException {
+        try (
+            PreparedStatement query = connection.prepareStatement(rawSqlQuery);
+            ResultSet results = query.executeQuery()
+        ) {
+            results.next();
+            return results.getLong("lsn");
```

Comment:
I remember we had to parse it in Postgres. Aurora seems to be more straightforward.

Reply:
Exactly, in postgres LSN has text format like ''16/B374D848". Aurora returns just long type.